0001-dinerojs-v2.md

```diff
@@ -0,0 +1,188 @@
+
+It should be organized in a monorepo to make it easier to create companion services and factories.
+
+### 📐 API
+
+Dinero.js v2 should expose the following:
+
+- The core API, similar to v1. It should exclude deprecated v1 methods such as [`hasCents`](https://sarahdayan.github.io/dinero.js/module-Dinero.html#~hasCents) and fix design issues such as the dependency on the I18n API ([`toFormat`](https://sarahdayan.github.io/dinero.js/module-Dinero.html#~toFormat)) or on a REST API ([`convert`](https://sarahdayan.github.io/dinero.js/module-Dinero.html#~convert)).
+- An optional service which allows currency assertions and automatic exponent inclusion.
+- A set of helpers to create specific `Dinero` objects more easily.
+
+#### Core API
+
+##### Current design flaws
+
+The main flaws in Dinero.js v1 are how it encapsulates certain behaviors, locking users into a specific usage, and not allowing for flexibility.
+
+Two good examples are `toFormat` and `convert`:
+
+- `toFormat` wraps around the I18n API. It creates two issues:
+  - The API doesn't output an exact desired format, but a localized currency string. This behavior is confusing, because Dinero lets users pass a mask, so they reasonably expect an exact projection.
+  - The I18n API has inconsistent implementations across different environments, which makes the API unpredictable and even more confusing.
+- `convert` wraps around a provided REST API to hide the request logic. It creates two issues:
+  - It forces users to consume a REST API, preventing them from querying rates from another source (in memory, a file) or use a different protocol (SOAP, GraphQL).
+  - It performs a new network request at each call, preventing users from caching results.
+
+`toFormat` should no longer rely on the I18n API, and instead format the object exactly as specified in the mask. The symbol would either be replaced by the currency in full or come from using the currency service. Users would still be able to use the I18n API through the usage of [`toRoundedUnit`](https://sarahdayan.github.io/dinero.js/module-Dinero.html#~toRoundedUnit).
+
+`convert` should no longer expect a REST API and encapsulate the fetching behavior. Instead, it should expect a Promise which resolves to a `Rates` object.
+
+```js
+const rates = new Promise(resolve =>
+  resolve({
+    EUR: 0.81162
+  })
+);
+
+Dinero({ amount: 500 }).convert("EUR", { rates });
+```
+
+##### Integers & BigInts
+
+Dinero.js v1 uses integers for `amount` (through the `number` primitive), and this shouldn't change, but Dinero.js v2 should also allow for `bigint` [as soon as the proposal goes to stage 4](https://tc39.es/proposal-bigint/).
+
+In the meantime, and to support environments with no `bigint` support, Dinero.js v2 should let users provide their own `Calculator`. This way, they can use whatever library they want and map the calculator functions. This behavior should be made available through a Dinero factory, so it doesn't "pollute" the API.
```

Comment:
At the interface between `Calculator` and `dinero`, would the number ever have to be coerced back to a Javascript `number`? That would break the big int guarantees as well.

Reply:
No, they wouldn't. You'd have the choice to use either `number`s or `bigint`s.

Dinero objects created with a `bigint` amount wouldn't be incompatible with Dinero objects created with a `number` amount, as you can't calculate both types together without coercing one first. Coercing to `number` would risk losing accuracy, while coercing to `bigint` would hurt performance.

There _could_ be a conversion of `number`s into `bigint`s though, via explicitly passing a parameter. 

```js
const d1 = Dinero({ amount: 20n })
const d2 = Dinero({ amount: 20 })

d1.add(d2) // TypeError: can't convert BigInt to number
d1.add(d2, true) // Dinero object with amount: 40n
```

What do you think?

Reply:
Hmmm I was thinking for the example you gave or legacy support using an arbitrary precision library like “big.js” by making a calculator object for it, that if dinero sets an underlying implementation for the actual data like number, then it couldn’t use the number format of the library (in the case of big, an object with multiple fields that preserves arbitrary precision); so I think for calculator to be useful, it should also have the power to dictate the underlying number format

Reply:
Yes indeed, that's why you would have access to a `createDinero` utility to which you could give your own calculator. Does that make sense?

Reply:
Yup! Just wasn’t clear from the reply above - not just the choice between number and bigint, but also potentially a totally different numeric data structure you can specify. So long as that’s the design of the calculator/createDinero utility it remains usable